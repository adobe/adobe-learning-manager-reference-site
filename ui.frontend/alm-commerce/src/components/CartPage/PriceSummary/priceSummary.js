/*******************************************************************************
 *
 *    Copyright 2021 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
//import { useConfigContext } from "@adobe/aem-core-cif-react-components";
import { gql } from "@apollo/client";
import { PriceSummaryFragment } from "@magento/peregrine/lib/talons/CartPage/PriceSummary/priceSummaryFragments.gql";
import { usePriceSummary } from "@magento/peregrine/lib/talons/CartPage/PriceSummary/usePriceSummary";
import { useStyle } from "@magento/venia-ui/lib/classify";
import Button from "@magento/venia-ui/lib/components/Button";
import DiscountSummary from "@magento/venia-ui/lib/components/CartPage/PriceSummary/discountSummary";
import GiftCardSummary from "@magento/venia-ui/lib/components/CartPage/PriceSummary/giftCardSummary";
import defaultClasses from "@magento/venia-ui/lib/components/CartPage/PriceSummary/priceSummary.module.css";
import ShippingSummary from "@magento/venia-ui/lib/components/CartPage/PriceSummary/shippingSummary";
import TaxSummary from "@magento/venia-ui/lib/components/CartPage/PriceSummary/taxSummary";
import Price from "@magento/venia-ui/lib/components/Price";
import React from "react";
import { FormattedMessage, useIntl } from "react-intl";
const GET_PRICE_SUMMARY = gql`
  query getPriceSummary($cartId: String!) {
    cart(cart_id: $cartId) {
      id
      ...PriceSummaryFragment
    }
  }
  ${PriceSummaryFragment}
`;

/**
 * A child component of the CartPage component.
 * This component fetches and renders cart data, such as subtotal, discounts applied,
 * gift cards applied, tax, shipping, and cart total.
 *
 * @param {Object} props
 * @param {Object} props.classes CSS className overrides.
 * See [priceSummary.css]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/CartPage/PriceSummary/priceSummary.css}
 * for a list of classes you can override.
 *
 * @returns {React.Element}
 *
 * @example <caption>Importing into your project</caption>
 * import PriceSummary from "@magento/venia-ui/lib/components/CartPage/PriceSummary";
 */
const PriceSummary = (props) => {
  const { isUpdating } = props;
  const classes = useStyle(defaultClasses, props.classes);
  const talonProps = usePriceSummary({
    queries: {
      getPriceSummary: GET_PRICE_SUMMARY,
    },
  });

  const { hasError, hasItems, isCheckout, isLoading, flatData } = talonProps;
  const { formatMessage } = useIntl();

  const { pagePaths } = { pagePaths: "dummy" }; //useConfigContext();

  const handleProceedToCheckout = () => {
    window.location.href = pagePaths.checkoutPage;
  };

  if (hasError) {
    return (
      <div className={classes.root}>
        <span className={classes.errorText}>
          <FormattedMessage
            id={"priceSummary.errorText"}
            defaultMessage={
              "Something went wrong. Please refresh and try again."
            }
          />
        </span>
      </div>
    );
  } else if (!hasItems) {
    return null;
  }

  const { subtotal, total, discounts, giftCards, taxes, shipping } = flatData;

  const isPriceUpdating = isUpdating || isLoading;
  const priceClass = isPriceUpdating ? classes.priceUpdating : classes.price;
  const totalPriceClass = isPriceUpdating
    ? classes.priceUpdating
    : classes.totalPrice;

  const totalPriceLabel = isCheckout
    ? formatMessage({
        id: "priceSummary.total",
        defaultMessage: "Total",
      })
    : formatMessage({
        id: "priceSummary.estimatedTotal",
        defaultMessage: "Estimated Total",
      });

  const proceedToCheckoutButton = !isCheckout ? (
    <div className={classes.checkoutButton_container}>
      <Button
        disabled={isPriceUpdating}
        priority={"high"}
        onClick={handleProceedToCheckout}
      >
        <FormattedMessage
          id={"priceSummary.checkoutButton"}
          defaultMessage={"Proceed to Checkout"}
        />
      </Button>
    </div>
  ) : null;

  return (
    <div className={classes.root}>
      <div className={classes.lineItems}>
        <span className={classes.lineItemLabel}>
          <FormattedMessage
            id={"priceSummary.lineItemLabel"}
            defaultMessage={"Subtotal"}
          />
        </span>
        <span className={priceClass}>
          <Price value={subtotal.value} currencyCode={subtotal.currency} />
        </span>
        <DiscountSummary
          classes={{
            lineItemLabel: classes.lineItemLabel,
            price: priceClass,
          }}
          data={discounts}
        />
        <GiftCardSummary
          classes={{
            lineItemLabel: classes.lineItemLabel,
            price: priceClass,
          }}
          data={giftCards}
        />
        <TaxSummary
          classes={{
            lineItemLabel: classes.lineItemLabel,
            price: priceClass,
          }}
          data={taxes}
          isCheckout={isCheckout}
        />
        <ShippingSummary
          classes={{
            lineItemLabel: classes.lineItemLabel,
            price: priceClass,
          }}
          data={shipping}
          isCheckout={isCheckout}
        />
        <span className={classes.totalLabel}>{totalPriceLabel}</span>
        <span className={totalPriceClass}>
          <Price value={total.value} currencyCode={total.currency} />
        </span>
      </div>
      {proceedToCheckoutButton}
    </div>
  );
};

export default PriceSummary;
