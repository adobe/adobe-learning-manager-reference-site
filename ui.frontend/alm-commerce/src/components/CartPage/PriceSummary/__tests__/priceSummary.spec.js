/*******************************************************************************
 *
 *    Copyright 2021 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { usePriceSummary } from '@magento/peregrine/lib/talons/CartPage/PriceSummary/usePriceSummary';
import PriceSummary from '../priceSummary';

import render from '../../../utils/test-utils';

jest.mock('@adobe/aem-core-cif-react-components', () => ({
    useConfigContext: jest.fn(() => ({
        pagePaths: {
            checkoutPage: '/content/venia/us/en/checkout.html'
        }
    }))
}));
jest.mock('react-router-dom', () => ({
    Link: jest.fn(() => 'Proceed to Checkout')
}));
jest.mock('@apollo/client', () => {
    const runQuery = jest.fn();
    const queryResult = {
        data: {
            cart: {
                items: [
                    {
                        quantity: 1
                    }
                ],
                applied_gift_cards: [],
                shipping_addresses: [
                    {
                        selected_shipping_method: {
                            amount: {
                                value: 0,
                                currency: 'USD'
                            }
                        }
                    }
                ],
                prices: {
                    subtotal_excluding_tax: {
                        currency: 'USD',
                        value: 11
                    },
                    grand_total: {
                        currency: 'USD',
                        value: 10
                    },
                    discounts: [
                        {
                            amount: {
                                value: 1,
                                currency: 'USD'
                            }
                        }
                    ],
                    applied_taxes: [
                        {
                            amount: {
                                value: 0,
                                currency: 'USD'
                            }
                        }
                    ]
                }
            }
        },
        error: null,
        loading: false
    };
    const useLazyQuery = jest.fn(() => [runQuery, queryResult]);

    return {
        gql: jest.fn(),
        useLazyQuery
    };
});

const defaultTalonProps = {
    hasError: false,
    hasItems: true,
    isCheckout: false,
    isLoading: false,
    flatData: {
        subtotal: { currency: 'USD', value: 3.5 },
        total: { currency: 'USD', value: 8.5 },
        discounts: null,
        giftCards: [],
        taxes: [],
        shipping: { currency: 'USD', value: 5 }
    }
};

jest.mock('@magento/peregrine/lib/talons/CartPage/PriceSummary/usePriceSummary', () => ({
    usePriceSummary: jest.fn().mockReturnValue({
        hasError: false,
        hasItems: true,
        isCheckout: false,
        isLoading: false,
        flatData: {
            subtotal: { currency: 'USD', value: 3.5 },
            total: { currency: 'USD', value: 8.5 },
            discounts: null,
            giftCards: [],
            taxes: [],
            shipping: { currency: 'USD', value: 5 }
        }
    })
}));

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = { cartId: 'cart123' };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

test('renders PriceSummary correctly on cart page', () => {
    const tree = render(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders PriceSummary correctly on checkout page', () => {
    usePriceSummary.mockReturnValueOnce({
        ...defaultTalonProps,
        isCheckout: true
    });

    const tree = render(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders an error state if query fails', () => {
    usePriceSummary.mockReturnValueOnce({
        ...defaultTalonProps,
        hasError: true
    });

    const tree = render(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders summary with loading state if query is loading', () => {
    usePriceSummary.mockReturnValueOnce({
        ...defaultTalonProps,
        isLoading: true
    });

    const tree = render(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders nothing if query returns no items', () => {
    usePriceSummary.mockReturnValueOnce({
        ...defaultTalonProps,
        hasItems: false
    });

    const tree = render(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});
