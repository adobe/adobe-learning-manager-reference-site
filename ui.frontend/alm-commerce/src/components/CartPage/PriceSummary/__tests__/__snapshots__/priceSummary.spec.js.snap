// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PriceSummary correctly on cart page 1`] = `
<div>
  <div>
    <span>
      Subtotal
    </span>
    <span>
      <span>
        $
      </span>
      <span>
        3
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
    <span>
      Estimated Total
    </span>
    <span>
      <span>
        $
      </span>
      <span>
        8
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
  </div>
  <div>
    <button
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span>
        Proceed to Checkout
      </span>
    </button>
  </div>
</div>
`;

exports[`renders PriceSummary correctly on checkout page 1`] = `
<div>
  <div>
    <span>
      Subtotal
    </span>
    <span>
      <span>
        $
      </span>
      <span>
        3
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
    <span>
      Total
    </span>
    <span>
      <span>
        $
      </span>
      <span>
        8
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
  </div>
</div>
`;

exports[`renders an error state if query fails 1`] = `
<div>
  <span>
    Something went wrong. Please refresh and try again.
  </span>
</div>
`;

exports[`renders nothing if query returns no items 1`] = `null`;

exports[`renders summary with loading state if query is loading 1`] = `
<div>
  <div>
    <span>
      Subtotal
    </span>
    <span>
      <span>
        $
      </span>
      <span>
        3
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
    <span>
      Estimated Total
    </span>
    <span>
      <span>
        $
      </span>
      <span>
        8
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
  </div>
  <div>
    <button
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span>
        Proceed to Checkout
      </span>
    </button>
  </div>
</div>
`;
